src/connection.cpp

```diff
@@ -514,119 +395,172 @@ namespace oxen::quic
+            uint32_t flags = FLAG_WRITE_MORE;
+            int64_t stream_id;
+            auto* source = channels.front();
+            channels.pop_front();  // Pop it off; if this stream should be checked again, append just
+                                   // before streams_end_it.
 
-            assert(stream || strs.empty());  // We should only get -1 at the end of the list
+            std::vector<ngtcp2_vec> bufs = source->pending();
 
-            const int64_t stream_id = stream ? stream->_stream_id : -1;
-            std::vector<ngtcp2_vec> bufs;
-            if (stream)
+            // this block will execute all "real" streams plus the "pseudo stream" of ID -1 to finish
+            // off any packets that need to be sent
+            if (source->is_stream())
             {
-                bufs = stream->pending();
-
-                if (stream->is_closing && !stream->sent_fin && stream->unsent() == 0)
+                stream_id = source->stream_id();
+                if (source->is_closing() && !source->sent_fin() && source->unsent() == 0)
                 {
                     log::trace(log_cat, "Sending FIN");
                     flags |= NGTCP2_WRITE_STREAM_FLAG_FIN;
-                    stream->sent_fin = true;
+                    source->set_fin(true);
                 }
-                else if (bufs.empty())
+                else if (stream_id != -1 && bufs.empty())  // we cant continue here with pseudo-stream, so check ID
                 {
                     log::debug(log_cat, "pending() returned empty buffer for stream ID {}, moving on", stream_id);
                     continue;
                 }
-            }
 
-            ngtcp2_ssize ndatalen;
-            auto nwrite = ngtcp2_conn_writev_stream(
-                    conn.get(),
-                    _path,
-                    &pkt_info,
-                    buf_pos,
-                    MAX_PMTUD_UDP_PAYLOAD,
-                    &ndatalen,
-                    flags,
-                    stream_id,
-                    bufs.data(),
-                    bufs.size(),
-                    ts);
-
-            log::trace(log_cat, "add_stream_data for stream {} returned [{},{}]", stream_id, nwrite, ndatalen);
-
-            if (nwrite < 0)
-            {
-                if (nwrite == NGTCP2_ERR_WRITE_MORE)  // -240
+                nwrite = ngtcp2_conn_writev_stream(
+                        conn.get(),
+                        _path,
+                        &pkt_info,
+                        buf_pos,
+                        MAX_PMTUD_UDP_PAYLOAD,
+                        &ndatalen,
+                        flags,
+                        stream_id,
+                        bufs.data(),
+                        bufs.size(),
+                        ts);
+
+                log::trace(log_cat, "add_stream_data for stream {} returned [{},{}]", stream_id, nwrite, ndatalen);
+                // ndatalen "should" be >0 for streams and <0 for psuedo streams
```

Comment:
The docs seem sorta ambiguous, but it *looks* like for the `stream id == -1` case, ndatalen will be either -1 or 0, so not necessarily negative?

Reply:
I was unsure, i never saw a ndatalen of 0 for the stream id of -1, and like you said the docs were ambiguous. ill look into it :+1: